### PR TITLE
Server WebSocket Skeleton extended with tests

### PR DESCRIPTION
**Pull request by MJHC**
- tests
 - some things have not been made because of Account is unfinshed
 - envelope message sending is now possible (not tested yet because of account)
- skeleton for different endpoint handling added

### CODE REVIEW

File: common/src/web_api.rs [hunk 1]
```diff
@@ -385,3 +387,20 @@ impl PreKeyResponseItem {
...
+    pub destination: uuid::Uuid,
+    pub timestamp: u64,
+    pub messages: Vec<SignalMessage>,
+    pub online: bool,
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+#[serde(rename_all = "camelCase")]
+pub struct SignalMessage {
+    pub r#type: u32,
```
Comment (Mikkel-Boje) [inline common/src/web_api.rs:402]: Can we change this name to something without special characters

File: server/client.py [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+    ssl_context = ssl.create_default_context(purpose=ssl.Purpose.SERVER_AUTH)
+    ssl_context.load_verify_locations(cafile="cert/rootCA.crt")
+    
+    async with websockets.connect(uri, ssl=ssl_context, ) as websocket: #extra_headers={"Authorization": sys.argv[1]}
+        await websocket.send("Hello from Python Client")
+        res = await websocket.recv()
+        print(res)
+
+# Run the event loop
+asyncio.get_event_loop().run_until_complete(connect())
```
Comment (Mikkel-Boje) [inline server/client.py]: FØJ PYTHON

File: server/src/socket.rs [hunk 3]
```diff
@@ -1,98 +1,526 @@
...
 
-    async fn add_ws(&self, who: SocketAddr, socket: WebSocket) {
-        let socket = Arc::new(Mutex::new(socket));
-        self.sockets.write().await.insert(who, socket);
+    async fn handle_request(
+        &mut self,
+        connection: &mut WebSocketConnection<T>,
+        req: WebSocketRequestMessage,
+    ) {
+        let uri = Uri::from_str(req.path()).unwrap();
```
Comment (Mikkel-Boje) [inline server/src/socket.rs:162]: unwrap nono

File: server/src/socket.rs [hunk 4]
```diff
@@ -1,98 +1,526 @@
...
-        self.ws_send(
-            &who,
-            Message::Binary(format!("hello {}", who).as_bytes().to_vec()),
-        )
-        .await;
+    async fn on_ws_binary(&mut self, connection: &mut WebSocketConnection<T>, bytes: Vec<u8>) {
+        let msg = match WebSocketMessage::decode(Bytes::from(bytes)) {
+            Ok(x) => x,
+            Err(y) => {
+                println!("handle_socket ws proto ERROR: {}", y);
```
Comment (Laursen79) [inline server/src/socket.rs:130]:
```suggestion
                println!("on_ws_binary could not decode bytes: {}", y);
```

File: server/src/socket.rs [hunk 5]
```diff
@@ -1,98 +1,526 @@
...
+        sender.send(Ok(Message::Close(None))).await;
+
+        assert!(
+            !state.lock().await.is_active(),
+            ">>>>>>>>>>>>>>>>>>>>> State was active, expected was closed"
+        )
+    }
+
+    #[tokio::test]
+    async fn test_text_relay_ok() {
```
Comment (Laursen79) [inline server/src/socket.rs:502]:
```suggestion
    async fn test_send_text_response_ok() {
```

**Review by Laursen79 — CHANGES_REQUESTED** [review]

**Review by Laursen79 — DISMISSED** [review]

**Review by Mikkel-Boje — CHANGES_REQUESTED** [review]
All around looking good with the logic. But some minor code cleanup

File: common/src/web_api.rs (hunk 1 above)
Comment (MJHC) [reply common/src/web_api.rs:402]: no, this is defined by signal

File: server/src/socket.rs [hunk 6]
```diff
@@ -1,98 +1,526 @@
...
-        )
-        .await;
+    async fn on_ws_binary(&mut self, connection: &mut WebSocketConnection<T>, bytes: Vec<u8>) {
+        let msg = match WebSocketMessage::decode(Bytes::from(bytes)) {
+            Ok(x) => x,
+            Err(y) => {
+                println!("handle_socket ws proto ERROR: {}", y);
+                return;
+            }
+        };
```
[the remark this answers is not in the document]
Comment (MJHC) [reply server/src/socket.rs]: does not work

**Review by Laursen79 — APPROVED** [review]

**Review by Mikkel-Boje — APPROVED** [review]

**Pull request merged**